### PR TITLE
manage schema cache when horizontally scaled (closes #1182)

### LINKED ISSUE #1182: improve schema management when horizontally scaled
Currently, when applying schema changes there should only be one instance running. We can get rid of this limitation by listening to changes on tables in `hdb_catalog` and reload the metadata when needed.

### PR DESCRIPTION
**Pull request by rakeshkky**
### Description

What component does this PR affect? 

- [x] Server
- [ ] Console
- [ ] CLI
- [ ] Docs
- [ ] Community Content
- [ ] Build System

Requires changes from other components? If yes, please mark the components:

- [ ] Server
- [ ] Console
- [ ] CLI
- [x] Docs
- [ ] Community Content
- [ ] Build System

### Related Issue

close #1182 

### Solution and Design

1. Haskel library `pg-client-hs` has been updated to expose a function that helps listen to `postgres` notifications over a `channel` in this [PR](https://github.com/hasura/pg-client-hs/pull/5)
2. The server records an event in a table `hdb_catalog.hdb_cache_update_event` whenever any `/v1/query` (that changes metadata) is requested. A trigger notifies a `cache update` event via `hasura_cache_update` channel
3. The server runs two concurrent threads namely `listener` and `processor`. The `listener` thread listens to events on `hasura_cache_update` channel and pushed into a `Queue`. The `processor` thread fetches events from that `Queue` and processes it. Thus server rebuilds schema cache from database and updates.

### Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Docs update
- [ ] Community content

### Checklist:

- [x] I have read the **[contributing guide](https://github.com/hasura/graphql-engine/blob/master/CONTRIBUTING.md)** and my code conforms to the guidelines.
- [x] This change requires a change in the documentation. 
- [ ] I have updated the documentation accordingly.
- [ ] I have added required tests.

### CODE REVIEW

**Comment by coco98** [comment]
We should add tests for this? Will require some interesting black magic from @shahidhk @arvi3411301 I think ;)

File: server/src-lib/Hasura/Server/CacheUpdate.hs [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+      -- Push a listen failed event to queue
+      STM.atomically $ STM.writeTQueue eventsQueue CUEListenFail
+      logInfo logger threadType $
+        object ["message" .= ("retrying in 10 seconds" :: T.Text)]
+      C.threadDelay $ 10 * 1000 * 1000
+
+    -- Postgres notification handler
+    notifyHandler notif = do
+      let eventChannel = PGChannel $ bsToTxt $ PQ.notifyRelname notif
+      when (eventChannel == pgChannel) $
```
Comment (0x777) [inline server/src-lib/Hasura/Server/CacheUpdate.hs]: Why do this check again?

File: server/src-lib/Hasura/Server/CacheUpdate.hs [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+  forever $ do
+    listenResE <- liftIO $ runExceptT $ listen pool pgChannel notifyHandler
+    either onError return listenResE
+  where
+    threadType = TTListener
+
+    onError e = do
+      logError logger threadType $ TEQueryError e
+      -- Push a listen failed event to queue
+      STM.atomically $ STM.writeTQueue eventsQueue CUEListenFail
```
Comment (0x777) [inline server/src-lib/Hasura/Server/CacheUpdate.hs]: Don't emit a 'listen error' event. Emit an event once the connection is re-established after an error. You'll probably need to expose this functionality from pg-client-hs, probably `listen` can take something like this as an argument:
```haskell
data NotifyHandler 
  = NotifyHandler 
  { _ehOnInit :: IO ()
  , _ehOnMessage :: Notify -> IO ()
  }
```

File: server/src-lib/Hasura/Server/CacheUpdate.hs [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+  -> STM.TQueue CacheUpdateEvent
+  -> IORef.IORef SchemaCache
+  -> MVar ()
+  -> UUID
+  -> UTC.UTCTime
+  -> IO ()
+cacheUpdateEventProcessor pool logger httpManager eventsQueue
+                       cacheRef lk serverId cacheInit = do
+  -- Initiate previous event IO reference
+  prevEventRef <- IORef.newIORef Nothing
```
Comment (0x777) [inline server/src-lib/Hasura/Server/CacheUpdate.hs]: You don't have to store the previous event ref after the above suggested change (emit event on reconnection).

**Pull request merged**